compiler/rustc_lint/src/builtin.rs

```diff
@@ -1458,15 +1458,19 @@ impl TypeAliasBounds {
 
 impl<'tcx> LateLintPass<'tcx> for TypeAliasBounds {
     fn check_item(&mut self, cx: &LateContext<'_>, item: &hir::Item<'_>) {
-        let hir::ItemKind::TyAlias(ty, type_alias_generics) = &item.kind else { return };
-        if cx.tcx.type_of(item.owner_id.def_id).skip_binder().has_opaque_types() {
-            // Bounds are respected for `type X = impl Trait` and `type X = (impl Trait, Y);`
-            return;
-        }
-        if cx.tcx.type_of(item.owner_id).skip_binder().has_inherent_projections() {
-            // Bounds are respected for `type X = … Type::Inherent …`
+        let hir::ItemKind::TyAlias(hir_ty, type_alias_generics) = &item.kind else { return };
+
+        let ty = cx.tcx.type_of(item.owner_id).skip_binder();
+        if cx.tcx.features().lazy_type_alias
+            || ty.has_opaque_types()
+            || ty.has_inherent_projections()
+        {
+            // Bounds of lazy type aliases and on eager ones that contain opaque types or inherent
```

Comment:
Let me know if the phrasing is ok, *bounds* sounds ambiguous to me. One might wonder it refers to the bounds on the generic parameters (which is the case) or if it refers to the *item bounds* (i.e. the `B` in `type T: B = …` which is obviously invalid for type aliases).